### PR TITLE
bpo-45711: Do not push/pop traceback on stack as part of exc_info

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue45711

### CODE REVIEW

**Comment by iritkatriel** [comment]
Marked as draft because I still need to update the opcodes documentation

File: Python/ceval.c [hunk 1]
```diff
@@ -2782,7 +2792,14 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             traceback = exc_info->exc_traceback;
             exc_info->exc_type = POP();
             exc_info->exc_value = POP();
-            exc_info->exc_traceback = POP();
+
+            assert(exc_info->exc_value);
+            if (exc_info->exc_value != Py_None) {
+                exc_info->exc_traceback = PyException_GetTraceback(exc_info->exc_value);
```
Comment (iritkatriel) [inline Python/ceval.c:2777]: The next step is to remove exc_traceback from exc_info and then this will not be here.

**Comment by iritkatriel** [comment]
This clashes with the except* branch, and I'd rather merge that one first and update this one.

**Comment by markshannon** [comment]
> This clashes with the except* branch, and I'd rather merge that one first and update this one.

I would rather merge this first.
Apart from the trivial conflict in dis.rst this seems ready to merge. The except* branch is not.

**Comment by iritkatriel** [comment]
> > This clashes with the except* branch, and I'd rather merge that one first and update this one.
> 
> I would rather merge this first. Apart from the trivial conflict in dis.rst this seems ready to merge. The except* branch is not.

We won't see a speedup for this because of the get_exc_traceback calls. We should remove them before measuring, and probably merge the whole thing together.

**Comment by iritkatriel** [comment]
> We won't see a speedup for this because of the get_exc_traceback calls. We should remove them before measuring, and probably merge the whole thing together.

And that requires PR 29922.

File: Python/compile.c [hunk 2]
```diff
@@ -1049,9 +1050,9 @@ stack_effect(int opcode, int oparg, int jump)
         case POP_BLOCK:
             return 0;
         case POP_EXCEPT:
-            return -3;
+            return -PY_EXC_INFO_STACK_SIZE;
         case POP_EXCEPT_AND_RERAISE:
-            return -7;
+            return -1 -2 * PY_EXC_INFO_STACK_SIZE;
```
Comment (arhadthedev) [inline Python/compile.c]:
```suggestion
            return -1 - 2 * PY_EXC_INFO_STACK_SIZE;
```

or even

```suggestion
            return -(1 + 2 * PY_EXC_INFO_STACK_SIZE);
```

File: Python/ceval.c [hunk 3]
```diff
@@ -2759,7 +2767,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
         }
 
         TARGET(POP_EXCEPT_AND_RERAISE) {
-            PyObject *lasti = PEEK(4);
+            PyObject *lasti = PEEK(PY_EXC_INFO_STACK_SIZE+1);
```
Comment (arhadthedev) [inline Python/ceval.c]:
```suggestion
            PyObject *lasti = PEEK(PY_EXC_INFO_STACK_SIZE + 1);
```

File: Python/ceval.c [hunk 4]
```diff
@@ -2792,7 +2809,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
 
         TARGET(RERAISE) {
             if (oparg) {
-                PyObject *lasti = PEEK(oparg+3);
+                PyObject *lasti = PEEK(oparg+PY_EXC_INFO_STACK_SIZE);
```
Comment (arhadthedev) [inline Python/ceval.c]:
```suggestion
                PyObject *lasti = PEEK(oparg + PY_EXC_INFO_STACK_SIZE);
```

**Review by arhadthedev — CHANGES_REQUESTED** [review]
Spacing between tokens is strange.

**Comment by iritkatriel** [comment]
Superseded by https://github.com/python/cpython/pull/30122

**Pull request closed without merging**